File: trading-engine/src/trading_engine/portfolio_calc.py
"""目標ポートフォリオ計算 → 注文生成"""

from dataclasses import dataclass

import sys

sys.path.insert(0, str(__import__("pathlib").Path(__file__).resolve().parents[3] / "mcp-server" / "src"))
from claude_trade.models import Signal  # noqa: E402


@dataclass
class OrderRequest:
    symbol: str
    side: str  # BUY or SELL
    quantity: int
    signal_id: str
    reasoning: str
    exchange: str = "SMART"
    currency: str = "USD"


def calculate_target_shares(
    signal: Signal,
    nav: float,
    current_position: int,
    current_price: float,
    max_position_pct: float,
) -> int:
    """シグナルから目標株数を計算する

    - buy シグナル: strength * confidence に比例して NAV の一定割合を配分
    - sell シグナル: ポジションを縮小/クローズ
    - hold / avoid: 変更なし
    """
    if signal.signal_type in ("hold", "avoid"):
        return current_position

    max_position_value = nav * (max_position_pct / 100.0)

    if signal.signal_type == "buy":
        # strength * confidence で配分率を決定 (0〜max_position_pct)
        alloc_factor = abs(signal.strength) * signal.confidence
        target_value = max_position_value * alloc_factor
        target_shares = int(target_value / current_price) if current_price > 0 else 0
        return target_shares

    if signal.signal_type == "sell":
        # sell の場合: strength に応じてポジションを縮小
        reduce_factor = abs(signal.strength) * signal.confidence
        target_shares = int(current_position * (1 - reduce_factor))
        return max(0, target_shares)

    return current_position


def generate_orders(
    signals: dict[str, Signal],
    nav: float,
    positions: dict[str, int],  # symbol -> current quantity
    prices: dict[str, float],  # symbol -> current price
    max_position_pct: float,
    exchange: str = "SMART",
    currency: str = "USD",
) -> list[OrderRequest]:
    """シグナル群から注文リストを生成する"""
    orders: list[OrderRequest] = []

    for symbol, signal in signals.items():
        current_qty = positions.get(symbol, 0)
        current_price = prices.get(symbol, 0.0)

        if current_price <= 0:
            continue

        target_qty = calculate_target_shares(
            signal, nav, current_qty, current_price, max_position_pct
        )

        diff = target_qty - current_qty

        if diff > 0:
            orders.append(OrderRequest(
                symbol=symbol,
                side="BUY",
                quantity=diff,
                signal_id=str(signal.id),
                reasoning=f"Signal: {signal.signal_type} strength={signal.strength:.2f} conf={signal.confidence:.2f}",
                exchange=exchange,
                currency=currency,
            ))
        elif diff < 0:
            orders.append(OrderRequest(
                symbol=symbol,
                side="SELL",
                quantity=abs(diff),
                signal_id=str(signal.id),
                reasoning=f"Signal: {signal.signal_type} strength={signal.strength:.2f} conf={signal.confidence:.2f}",
                exchange=exchange,
                currency=currency,
            ))

    return orders

File: packages/server/src/trading-engine/portfolio-calc.ts
/** 目標ポートフォリオ計算 → 注文生成 */

import type { Signal } from "../db/schema.js";

export interface OrderRequest {
  symbol: string;
  side: "BUY" | "SELL";
  quantity: number;
  signalId: string;
  reasoning: string;
  exchange: string;
  currency: string;
}

function calculateTargetShares(
  signal: Signal,
  nav: number,
  currentPosition: number,
  currentPrice: number,
  maxPositionPct: number,
): number {
  if (signal.signalType === "hold" || signal.signalType === "avoid") {
    return currentPosition;
  }

  const maxPositionValue = nav * (maxPositionPct / 100);

  if (signal.signalType === "buy") {
    const allocFactor = Math.abs(signal.strength) * (signal.confidence ?? 0.5);
    const targetValue = maxPositionValue * allocFactor;
    return currentPrice > 0 ? Math.floor(targetValue / currentPrice) : 0;
  }

  if (signal.signalType === "sell") {
    const reduceFactor = Math.abs(signal.strength) * (signal.confidence ?? 0.5);
    return Math.max(0, Math.floor(currentPosition * (1 - reduceFactor)));
  }

  return currentPosition;
}

export function generateOrders(
  signalMap: Map<string, Signal>,
  nav: number,
  positions: Map<string, number>,
  prices: Map<string, number>,
  maxPositionPct: number,
  exchange = "SMART",
  currency = "USD",
): OrderRequest[] {
  const orders: OrderRequest[] = [];

  for (const [symbol, signal] of signalMap) {
    const currentQty = positions.get(symbol) ?? 0;
    const currentPrice = prices.get(symbol) ?? 0;

    if (currentPrice <= 0) continue;

    const targetQty = calculateTargetShares(
      signal,
      nav,
      currentQty,
      currentPrice,
      maxPositionPct,
    );

    const diff = targetQty - currentQty;

    if (diff > 0) {
      orders.push({
        symbol,
        side: "BUY",
        quantity: diff,
        signalId: signal.id,
        reasoning: `Signal: ${signal.signalType} strength=${signal.strength.toFixed(2)} conf=${(signal.confidence ?? 0.5).toFixed(2)}`,
        exchange,
        currency,
      });
    } else if (diff < 0) {
      orders.push({
        symbol,
        side: "SELL",
        quantity: Math.abs(diff),
        signalId: signal.id,
        reasoning: `Signal: ${signal.signalType} strength=${signal.strength.toFixed(2)} conf=${(signal.confidence ?? 0.5).toFixed(2)}`,
        exchange,
        currency,
      });
    }
  }

  return orders;
}
